### PR TITLE
Sort callbacks so that CheckpointSaver goes before HuggingFaceCheckpointer

### PR DESCRIPTION
**Pull request by irenedea**
Sort the callbacks so that CheckpointSaver goes first, then HuggingFaceCheckpointer, and finally all other callbacks. This saves time during final checkpoint saving and uploading.

A rough sketch on the reordering based on Llama 70B profiling:
Before
image

After
image

e2e run in staging: `ift-meta-llama-3-8b-aoprwt-reordered-q01HKT`

### CODE REVIEW

File: llmfoundry/command_utils/train.py [hunk 1]
```diff
@@ -187,6 +188,24 @@ def _initialize_dist_with_barrier(dist_timeout: Union[int, float]):
...
+        trainer (Trainer): Trainer object
+    """
+
+    def _sort_key(c: Callback) -> int:
+        # CheckpointSaver goes first because the blocking time is shortest while upload is async.
+        if isinstance(c, CheckpointSaver):
+            return 0
+        if isinstance(c, HuggingFaceCheckpointer):
+            return 1
+        return 2
```
Comment (mvpatel2000) [inline llmfoundry/command_utils/train.py:204]: should other callbacks be before or after ckpting? imo before as they might change model

**Review by mvpatel2000 — COMMENTED** [review]
test?

File: llmfoundry/command_utils/train.py (hunk 1 above)
Comment (irenedea) [reply llmfoundry/command_utils/train.py:204]: Ah good point, i'll switch to before

**Comment by mvpatel2000** [comment]
> @mvpatel2000 What tests would you like to see? I can add a test for the helper function that checks to see that the callbacks are in the expected order. Anything else?

I think that's fine!

**Review by mvpatel2000 — APPROVED** [review]

**Pull request merged**